A panel is a 3D container for data. The term data Panel is derived from econometrics and is partially responsible for the name Pandas.

A panel is a 3D container for data. The term data Panel is derived from econometrics and is partially responsible for the name Pandas.

### 1. Create a random 3D datasets

source: https://www.tutorialspoint.com/python_pandas/python_pandas_panel

In [136]:
import pandas as pd
import numpy as np

data = {'Item1' : pd.DataFrame(np.random.randn(4, 3)), 
   'Item2' : pd.DataFrame(np.random.randn(4, 2))}
p = pd.Panel(data)
p

C:\Users\SAM\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:2963: FutureWarning: 
Panel is deprecated and will be removed in a future version.
The recommended way to represent these types of 3-dimensional data are with a MultiIndex on a DataFrame, via the Panel.to_frame() method
Alternatively, you can use the xarray package http://xarray.pydata.org/en/stable/.
Pandas provides a `.to_xarray()` method to help automate this conversion.

  exec(code_obj, self.user_global_ns, self.user_ns)


<class 'pandas.core.panel.Panel'>
Dimensions: 2 (items) x 4 (major_axis) x 3 (minor_axis)
Items axis: Item1 to Item2
Major_axis axis: 0 to 3
Minor_axis axis: 0 to 2

In [137]:
p["Item1"]

0      1      2
0  0.331 -0.559 -1.052
1 -0.654 -1.856  0.077
2 -0.010 -0.660 -1.144
3 -0.071  0.930  1.039

In [138]:
p["Item2"]

0      1   2
0 -2.501  0.595 NaN
1 -0.160  0.712 NaN
2 -1.166 -0.215 NaN
3  1.442  0.784 NaN

### 2. Create a 3D Dataset from 2D Stocks Datasets

In [139]:
all_companies = pd.read_csv("data/all_stocks_5yr.csv", 
                            parse_dates=["date"], index_col="date")

Here we have stock prices for 500 companies. The Name column indicates the company’s name. For this tutorial, we will be using Facebook (FB), Amazon (AMZN) and Microsoft (MSFT) data.

In [140]:
all_companies.head()

open   high    low  close    volume Name
date                                                 
2013-02-08  15.07  15.12  14.63  14.75   8407500  AAL
2013-02-11  14.89  15.01  14.26  14.46   8882000  AAL
2013-02-12  14.45  14.51  14.10  14.27   8126000  AAL
2013-02-13  14.30  14.94  14.25  14.66  10259500  AAL
2013-02-14  14.94  14.96  13.16  13.99  31879900  AAL

In [141]:
FB = all_companies[all_companies["Name"] == "FB"].copy()
AMZN = all_companies[all_companies["Name"] == "AMZN"].copy()
MSFT = all_companies[all_companies["Name"] == "MSFT"].copy()

In [142]:
FB.drop("Name", axis='columns', inplace=True)
AMZN.drop("Name", axis='columns', inplace=True)
MSFT.drop("Name", axis='columns', inplace=True)

In [143]:
FB.head(3)

open   high    low   close    volume
date                                             
2013-02-08  28.89  29.17  28.51  28.545  37662614
2013-02-11  28.61  28.68  28.04  28.260  36979533
2013-02-12  27.67  28.16  27.10  27.370  93417215

In [144]:
AMZN.head(3)

open    high      low   close   volume
date                                                
2013-02-08  261.40  265.25  260.555  261.95  3879078
2013-02-11  263.20  263.25  256.600  257.21  3403403
2013-02-12  259.19  260.16  257.000  258.70  2938660

In [145]:
MSFT.head(3)

open   high    low  close    volume
date                                            
2013-02-08  27.35  27.71  27.31  27.55  33318306
2013-02-11  27.65  27.92  27.50  27.86  32247549
2013-02-12  27.88  28.00  27.75  27.88  35990829

In [146]:
all_companies.Name.unique()

array(['AAL', 'AAPL', 'AAP', 'ABBV', 'ABC', 'ABT', 'ACN', 'ADBE', 'ADI',
       'ADM', 'ADP', 'ADSK', 'ADS', 'AEE', 'AEP', 'AES', 'AET', 'AFL',
       'AGN', 'AIG', 'AIV', 'AIZ', 'AJG', 'AKAM', 'ALB', 'ALGN', 'ALK',
       'ALLE', 'ALL', 'ALXN', 'AMAT', 'AMD', 'AME', 'AMGN', 'AMG', 'AMP',
       'AMT', 'AMZN', 'ANDV', 'ANSS', 'ANTM', 'AON', 'AOS', 'APA', 'APC',
       'APD', 'APH', 'APTV', 'ARE', 'ARNC', 'ATVI', 'AVB', 'AVGO', 'AVY',
       'AWK', 'AXP', 'AYI', 'AZO', 'A', 'BAC', 'BAX', 'BA', 'BBT', 'BBY',
       'BDX', 'BEN', 'BF.B', 'BHF', 'BHGE', 'BIIB', 'BK', 'BLK', 'BLL',
       'BMY', 'BRK.B', 'BSX', 'BWA', 'BXP', 'CAG', 'CAH', 'CAT', 'CA',
       'CBG', 'CBOE', 'CBS', 'CB', 'CCI', 'CCL', 'CDNS', 'CELG', 'CERN',
       'CFG', 'CF', 'CHD', 'CHK', 'CHRW', 'CHTR', 'CINF', 'CI', 'CLX',
       'CL', 'CMA', 'CMCSA', 'CME', 'CMG', 'CMI', 'CMS', 'CNC', 'CNP',
       'COF', 'COG', 'COL', 'COO', 'COP', 'COST', 'COTY', 'CPB', 'CRM',
       'CSCO', 'CSRA', 'CSX', 'CTAS', 'CTL', 'CTSH', 'CTXS

In [184]:
all_companies.Name.nunique()

505

In [185]:
data = {
    "FB" : FB,
    "AMZN" : AMZN,
    "MSFT": MSFT
}


In [186]:
p = pd.Panel(data)
p

C:\Users\SAM\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:2963: FutureWarning: 
Panel is deprecated and will be removed in a future version.
The recommended way to represent these types of 3-dimensional data are with a MultiIndex on a DataFrame, via the Panel.to_frame() method
Alternatively, you can use the xarray package http://xarray.pydata.org/en/stable/.
Pandas provides a `.to_xarray()` method to help automate this conversion.

  exec(code_obj, self.user_global_ns, self.user_ns)


<class 'pandas.core.panel.Panel'>
Dimensions: 3 (items) x 1259 (major_axis) x 5 (minor_axis)
Items axis: FB to MSFT
Major_axis axis: 2013-02-08 00:00:00 to 2018-02-07 00:00:00
Minor_axis axis: open to volume

We have extracted data for Facebook, Amazon, and Microsft and passed the data into Pandas Panel. Now, lets explore on Panel object.

In [187]:
type(p)

pandas.core.panel.Panel

In [188]:
p.items

Index(['FB', 'AMZN', 'MSFT'], dtype='object')

In [189]:
p.major_axis

DatetimeIndex(['2013-02-08', '2013-02-11', '2013-02-12', '2013-02-13',
               '2013-02-14', '2013-02-15', '2013-02-19', '2013-02-20',
               '2013-02-21', '2013-02-22',
               ...
               '2018-01-25', '2018-01-26', '2018-01-29', '2018-01-30',
               '2018-01-31', '2018-02-01', '2018-02-02', '2018-02-05',
               '2018-02-06', '2018-02-07'],
              dtype='datetime64[ns]', name='date', length=1259, freq=None)

In [190]:
p.minor_axis

Index(['open', 'high', 'low', 'close', 'volume'], dtype='object')

In [191]:
p.shape

(3, 1259, 5)

In [192]:
p.to_frame().head()

FB       AMZN       MSFT
date       minor                                  
2013-02-08 open    2.889e+01  2.614e+02  2.735e+01
           high    2.917e+01  2.652e+02  2.771e+01
           low     2.851e+01  2.606e+02  2.731e+01
           close   2.855e+01  2.619e+02  2.755e+01
           volume  3.766e+07  3.879e+06  3.332e+07

### 3. Extract a dataFrame from Panels

- The first argument in the bracket is the Item axis.
- The second argument in the bracket can be either the minor axis or major axis.

In [193]:
p

<class 'pandas.core.panel.Panel'>
Dimensions: 3 (items) x 1259 (major_axis) x 5 (minor_axis)
Items axis: FB to MSFT
Major_axis axis: 2013-02-08 00:00:00 to 2018-02-07 00:00:00
Minor_axis axis: open to volume

**The first bracket**

In [194]:
p["AMZN"].head()

open    high      low   close     volume
date                                                  
2013-02-08  261.40  265.25  260.555  261.95  3.879e+06
2013-02-11  263.20  263.25  256.600  257.21  3.403e+06
2013-02-12  259.19  260.16  257.000  258.70  2.939e+06
2013-02-13  261.53  269.96  260.300  269.47  5.293e+06
2013-02-14  267.37  270.65  265.400  269.24  3.463e+06

In [195]:
p["AMZN"].info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 1259 entries, 2013-02-08 to 2018-02-07
Data columns (total 5 columns):
open      1259 non-null float64
high      1259 non-null float64
low       1259 non-null float64
close     1259 non-null float64
volume    1259 non-null float64
dtypes: float64(5)
memory usage: 99.0 KB


To get a specific data, 3 arguments are needed. The sequence is as follows:

**[items, major , minor]**

In [196]:
p["AMZN"].head()

open    high      low   close     volume
date                                                  
2013-02-08  261.40  265.25  260.555  261.95  3.879e+06
2013-02-11  263.20  263.25  256.600  257.21  3.403e+06
2013-02-12  259.19  260.16  257.000  258.70  2.939e+06
2013-02-13  261.53  269.96  260.300  269.47  5.293e+06
2013-02-14  267.37  270.65  265.400  269.24  3.463e+06

In [197]:
p["AMZN","2013-02-08"]

open      2.614e+02
high      2.652e+02
low       2.606e+02
close     2.619e+02
volume    3.879e+06
Name: 2013-02-08 00:00:00, dtype: float64

In [198]:
p["AMZN","2013-02-08", "open"]

261.4

In [199]:
p["AMZN"]["volume"].describe()

count    1.259e+03
mean     3.730e+06
std      2.167e+06
min      1.093e+06
25%      2.511e+06
50%      3.145e+06
75%      4.220e+06
max      2.386e+07
Name: volume, dtype: float64

In [200]:
p["AMZN"]["volume"].min()

1092970.0

### 4. Set options on floating points

Change all the precision points values for floating points.

In [222]:
p["AMZN"].head(3)

open    high      low   close     volume
date                                                  
2013-02-08  261.40  265.25  260.555  261.95  3.879e+06
2013-02-11  263.20  263.25  256.600  257.21  3.403e+06
2013-02-12  259.19  260.16  257.000  258.70  2.939e+06

In [227]:
pd.set_option("precision", 1)

In [228]:
p["AMZN"].head(3)

open   high    low  close   volume
date                                           
2013-02-08  261.4  265.2  260.6  261.9  3.9e+06
2013-02-11  263.2  263.2  256.6  257.2  3.4e+06
2013-02-12  259.2  260.2  257.0  258.7  2.9e+06

In [229]:
pd.set_option("precision", 3)

### 5. Extract with loc[  ] , iloc [  ]  and ix[  ]  methods

**ITEMS:** Index(['FB', 'AMZN', 'MSFT'], dtype='object')

**MAJOR** : DatetimeIndex(['2013-02-08', '2013-02-11', '2013-02-12', '2013-02-13',
               '2013-02-14', '2013-02-15', '2013-02-19', '2013-02-20',
               '2013-02-21', '2013-02-22',
               ...
               '2018-01-25', '2018-01-26', '2018-01-29', '2018-01-30',
               '2018-01-31', '2018-02-01', '2018-02-02', '2018-02-05',
               '2018-02-06', '2018-02-07'],
               
              dtype='datetime64[ns]', name='date', length=1259, freq=None)
              
**MINOR** : Index(['open', 'high', 'low', 'close', 'volume'], dtype='object')

**[  items , major , minor ]**

This is the reference when you want to use loc[ ], iloc[ ] and ix[ ] methods

In [231]:
p

<class 'pandas.core.panel.Panel'>
Dimensions: 3 (items) x 1259 (major_axis) x 5 (minor_axis)
Items axis: FB to MSFT
Major_axis axis: 2013-02-08 00:00:00 to 2018-02-07 00:00:00
Minor_axis axis: open to volume

In [205]:
p.loc["FB",'2013-02-08' , "open"]

28.89

In [206]:
p.iloc[0,0,0]

28.89

In [207]:
p.ix["FB", 0 , "open"]

C:\Users\SAM\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
  """Entry point for launching an IPython kernel.


28.89

### 6. Convert Panel into MultiIndex DataFrame (and vice versa )

In [208]:
p

<class 'pandas.core.panel.Panel'>
Dimensions: 3 (items) x 1259 (major_axis) x 5 (minor_axis)
Items axis: FB to MSFT
Major_axis axis: 2013-02-08 00:00:00 to 2018-02-07 00:00:00
Minor_axis axis: open to volume

Major will be the first index and minor, the second index. The columns will be the Items list

In [209]:
p2 = p.to_frame()
p2

FB       AMZN       MSFT
date       minor                                  
2013-02-08 open    2.889e+01  2.614e+02  2.735e+01
           high    2.917e+01  2.652e+02  2.771e+01
           low     2.851e+01  2.606e+02  2.731e+01
           close   2.855e+01  2.619e+02  2.755e+01
           volume  3.766e+07  3.879e+06  3.332e+07
2013-02-11 open    2.861e+01  2.632e+02  2.765e+01
           high    2.868e+01  2.632e+02  2.792e+01
           low     2.804e+01  2.566e+02  2.750e+01
           close   2.826e+01  2.572e+02  2.786e+01
           volume  3.698e+07  3.403e+06  3.225e+07
2013-02-12 open    2.767e+01  2.592e+02  2.788e+01
           high    2.816e+01  2.602e+02  2.800e+01
           low     2.710e+01  2.570e+02  2.775e+01
           close   2.737e+01  2.587e+02  2.788e+01
           volume  9.342e+07  2.939e+06  3.599e+07
2013-02-13 open    2.736e+01  2.615e+02  2.793e+01
           high    2.832e+01  2.700e+02  2.811e+01
           low     2.731e+01  2.603e+02  2.788e+01
           close   2.791e+01  2.695e+02  2.803e+01
           volume  5.010e+07  5.293e+06  4.172e+07
2013-02-14 open    2.802e+01  2.674e+02  2.792e+01
           high    2.863e+01  2.706e+02  2.806e+01
           low     2.801e+01  2.654e+02  2.787e+01
           close   2.850e+01  2.692e+02  2.804e+01
           volume  3.558e+07  3.463e+06  3.266e+07
2013-02-15 open    2.852e+01  2.676e+02  2.804e+01
           high    2.875e+01  2.689e+02  2.816e+01
           low     2.809e+01  2.631e+02  2.788e+01
           close   2.832e+01  2.651e+02  2.801e+01
           volume  3.306e+07  3.980e+06  4.965e+07
...                      ...        ...        ...
2018-01-31 open    1.884e+02  1.451e+03  9.375e+01
           high    1.898e+02  1.473e+03  9.540e+01
           low     1.852e+02  1.450e+03  9.351e+01
           close   1.869e+02  1.451e+03  9.501e+01
           volume  4.328e+07  6.425e+06  4.876e+07
2018-02-01 open    1.882e+02  1.445e+03  9.479e+01
           high    1.953e+02  1.460e+03  9.607e+01
           low     1.879e+02  1.385e+03  9.358e+01
           close   1.931e+02  1.390e+03  9.426e+01
           volume  5.421e+07  9.114e+06  4.723e+07
2018-02-02 open    1.920e+02  1.477e+03  9.364e+01
           high    1.942e+02  1.498e+03  9.397e+01
           low     1.900e+02  1.414e+03  9.150e+01
           close   1.903e+02  1.430e+03  9.178e+01
           volume  2.668e+07  1.113e+07  4.787e+07
2018-02-05 open    1.869e+02  1.403e+03  9.056e+01
           high    1.906e+02  1.459e+03  9.324e+01
           low     1.806e+02  1.321e+03  8.800e+01
           close   1.813e+02  1.390e+03  8.800e+01
           volume  3.313e+07  1.149e+07  5.103e+07
2018-02-06 open    1.786e+02  1.361e+03  8.689e+01
           high    1.858e+02  1.444e+03  9.147e+01
           low     1.777e+02  1.352e+03  8.525e+01
           close   1.853e+02  1.443e+03  9.133e+01
           volume  3.776e+07  1.107e+07  6.800e+07
2018-02-07 open    1.842e+02  1.449e+03  9.049e+01
           high    1.851e+02  1.461e+03  9.177e+01
           low     1.799e+02  1.415e+03  8.920e+01
           close   1.802e+02  1.417e+03  8.961e+01
           volume  2.760e+07  7.163e+06  4.111e+07

[6295 rows x 3 columns]

**Multi Index to DataFrame**

In [210]:
p2.to_panel()

C:\Users\SAM\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: FutureWarning: 
Panel is deprecated and will be removed in a future version.
The recommended way to represent these types of 3-dimensional data are with a MultiIndex on a DataFrame, via the Panel.to_frame() method
Alternatively, you can use the xarray package http://xarray.pydata.org/en/stable/.
Pandas provides a `.to_xarray()` method to help automate this conversion.

  """Entry point for launching an IPython kernel.


<class 'pandas.core.panel.Panel'>
Dimensions: 3 (items) x 1259 (major_axis) x 5 (minor_axis)
Items axis: FB to MSFT
Major_axis axis: 2013-02-08 00:00:00 to 2018-02-07 00:00:00
Minor_axis axis: open to volume

### 7. .major_xs() method

- Takes the major axis data to retrieve data on Items and Minor axis.
- Returns only Items and Minor axis data. Major axis will be cut out.

In [211]:
p.major_axis

DatetimeIndex(['2013-02-08', '2013-02-11', '2013-02-12', '2013-02-13',
               '2013-02-14', '2013-02-15', '2013-02-19', '2013-02-20',
               '2013-02-21', '2013-02-22',
               ...
               '2018-01-25', '2018-01-26', '2018-01-29', '2018-01-30',
               '2018-01-31', '2018-02-01', '2018-02-02', '2018-02-05',
               '2018-02-06', '2018-02-07'],
              dtype='datetime64[ns]', name='date', length=1259, freq=None)

We can see the stock prices of every item (Facebook, Amazon, Microsoft) on 8th August 2012.


In [212]:
p.major_xs("2013-02-08")

FB       AMZN       MSFT
open    2.889e+01  2.614e+02  2.735e+01
high    2.917e+01  2.652e+02  2.771e+01
low     2.851e+01  2.606e+02  2.731e+01
close   2.855e+01  2.619e+02  2.755e+01
volume  3.766e+07  3.879e+06  3.332e+07

In [213]:
p.major_xs("2018-02-02")

FB       AMZN       MSFT
open    1.920e+02  1.477e+03  9.364e+01
high    1.942e+02  1.498e+03  9.397e+01
low     1.900e+02  1.414e+03  9.150e+01
close   1.903e+02  1.430e+03  9.178e+01
volume  2.668e+07  1.113e+07  4.787e+07

### 8. .minor_xs() method
- Returns the values of Major and Items
- Uses Minor as a key.
- Since we removed the minor axis, we are left with a 2-dimensional data. The value is returned in DataFrame.

In [214]:
p.minor_axis

Index(['open', 'high', 'low', 'close', 'volume'], dtype='object')

In [215]:
p.minor_xs("open")

FB     AMZN    MSFT
date                                
2013-02-08   28.890   261.40  27.350
2013-02-11   28.610   263.20  27.650
2013-02-12   27.670   259.19  27.880
2013-02-13   27.360   261.53  27.930
2013-02-14   28.020   267.37  27.920
2013-02-15   28.520   267.63  28.040
2013-02-19   28.230   265.91  27.880
2013-02-20   28.920   270.20  28.130
2013-02-21   28.280   265.12  27.740
2013-02-22   27.620   266.62  27.680
2013-02-25   27.160   266.94  27.970
2013-02-26   27.360   260.89  27.380
2013-02-27   27.340   259.40  27.420
2013-02-28   26.840   261.81  27.880
2013-03-01   27.050   263.27  27.720
2013-03-04   27.760   265.36  27.850
2013-03-05   27.880   274.00  28.290
2013-03-06   28.100   275.76  28.210
2013-03-07   27.570   274.10  28.110
2013-03-08   28.425   275.00  28.250
2013-03-11   28.010   273.43  27.940
2013-03-12   28.100   271.00  27.840
2013-03-13   27.620   275.24  27.870
2013-03-14   27.100   269.67  28.000
2013-03-15   27.030   264.98  28.030
2013-03-18   26.370   259.30  27.880
2013-03-19   26.530   258.40  28.120
2013-03-20   26.680   258.05  28.340
2013-03-21   25.660   256.11  28.110
2013-03-22   25.796   254.55  28.190
...             ...      ...     ...
2017-12-26  176.630  1168.36  85.310
2017-12-27  176.550  1179.91  85.650
2017-12-28  177.950  1189.00  85.900
2017-12-29  178.000  1182.35  85.630
2018-01-02  177.680  1172.00  86.125
2018-01-03  181.880  1188.30  86.055
2018-01-04  184.900  1205.00  86.590
2018-01-05  185.590  1217.51  87.660
2018-01-08  187.200  1236.00  88.200
2018-01-09  188.700  1256.90  88.650
2018-01-10  186.940  1245.15  87.860
2018-01-11  188.400  1259.74  88.130
2018-01-12  178.060  1273.39  88.670
2018-01-16  181.500  1323.00  90.100
2018-01-17  179.260  1312.24  89.080
2018-01-18  178.130  1293.95  89.800
2018-01-19  180.850  1312.00  90.140
2018-01-22  180.800  1297.17  90.000
2018-01-23  186.050  1338.09  91.900
2018-01-24  189.890  1374.82  92.550
2018-01-25  187.950  1368.00  92.465
2018-01-26  187.750  1392.01  93.120
2018-01-29  188.750  1409.18  95.140
2018-01-30  183.010  1403.17  93.300
2018-01-31  188.370  1451.30  93.750
2018-02-01  188.220  1445.00  94.790
2018-02-02  192.040  1477.39  93.640
2018-02-05  186.930  1402.62  90.560
2018-02-06  178.570  1361.46  86.890
2018-02-07  184.150  1449.00  90.490

[1259 rows x 3 columns]

### 9. .tranpose() method

We can swap the structure of the panel according to our needs. For example, swapping Items axis with Minor axis.

Lets look at our **current axis index **

- 0 - Items  axis
- 1 - Major axis
- 2 - Minor axis

In [216]:
p

<class 'pandas.core.panel.Panel'>
Dimensions: 3 (items) x 1259 (major_axis) x 5 (minor_axis)
Items axis: FB to MSFT
Major_axis axis: 2013-02-08 00:00:00 to 2018-02-07 00:00:00
Minor_axis axis: open to volume

The new structure of Panel that we want is:
- Item axis: open to volume
- Major axis: date
- Minor axis: company names ( FB to MSFT )

In [217]:
newPanel = p.transpose(2,1,0)
newPanel

C:\Users\SAM\Anaconda3\lib\site-packages\pandas\core\panel.py:1255: FutureWarning: 
Panel is deprecated and will be removed in a future version.
The recommended way to represent these types of 3-dimensional data are with a MultiIndex on a DataFrame, via the Panel.to_frame() method
Alternatively, you can use the xarray package http://xarray.pydata.org/en/stable/.
Pandas provides a `.to_xarray()` method to help automate this conversion.

  return super(Panel, self).transpose(*axes, **kwargs)


<class 'pandas.core.panel.Panel'>
Dimensions: 5 (items) x 1259 (major_axis) x 3 (minor_axis)
Items axis: open to volume
Major_axis axis: 2013-02-08 00:00:00 to 2018-02-07 00:00:00
Minor_axis axis: FB to MSFT

In [218]:
newPanel["open"].head()

FB    AMZN   MSFT
date                            
2013-02-08  28.89  261.40  27.35
2013-02-11  28.61  263.20  27.65
2013-02-12  27.67  259.19  27.88
2013-02-13  27.36  261.53  27.93
2013-02-14  28.02  267.37  27.92

In [219]:
newPanel.major_xs("2013-02-08")

open    high      low    close     volume
FB     28.89   29.17   28.510   28.545  3.766e+07
AMZN  261.40  265.25  260.555  261.950  3.879e+06
MSFT   27.35   27.71   27.310   27.550  3.332e+07

### 10. .swapaxes() method

In the transpose() method, we have to understand the index location of our axis. The swapaxes() method is a lot simpler. We can specify the 2 axes we want to swap.

In [220]:
p

<class 'pandas.core.panel.Panel'>
Dimensions: 3 (items) x 1259 (major_axis) x 5 (minor_axis)
Items axis: FB to MSFT
Major_axis axis: 2013-02-08 00:00:00 to 2018-02-07 00:00:00
Minor_axis axis: open to volume

We can get the same output as the transpose() method.

In [221]:
p.swapaxes("items", "minor")

C:\Users\SAM\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: FutureWarning: 
Panel is deprecated and will be removed in a future version.
The recommended way to represent these types of 3-dimensional data are with a MultiIndex on a DataFrame, via the Panel.to_frame() method
Alternatively, you can use the xarray package http://xarray.pydata.org/en/stable/.
Pandas provides a `.to_xarray()` method to help automate this conversion.

  """Entry point for launching an IPython kernel.


<class 'pandas.core.panel.Panel'>
Dimensions: 5 (items) x 1259 (major_axis) x 3 (minor_axis)
Items axis: open to volume
Major_axis axis: 2013-02-08 00:00:00 to 2018-02-07 00:00:00
Minor_axis axis: FB to MSFT